
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>2 BioFlows Tutorials &#8212; bioflows  documentation</title>
    <link rel="stylesheet" href="../_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <script type="text/javascript" src="../_static/documentation_options.js"></script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <link rel="author" title="About these documents" href="../about/" />
    <link rel="index" title="Index" href="../genindex/" />
    <link rel="search" title="Search" href="../search/" />
    <link rel="next" title="3 Wrappers" href="../wrappers/" />
    <link rel="prev" title="1 INTRODUCTION" href="../intro/" />
  
   
  <link rel="stylesheet" href="../_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  
  <link rel="alternate" type="application/atom+xml"  href="../blog/atom.xml" title="bioflows Blog">
  
  
  <link href="True" rel="stylesheet">
  
  <style type="text/css">
    ul.ablog-archive {list-style: none; overflow: auto; margin-left: 0px}
    ul.ablog-archive li {float: left; margin-right: 5px; font-size: 80%}
    ul.postlist a {font-style: italic;}
    ul.postlist-style-disc {list-style-type: disc;}
    ul.postlist-style-none {list-style-type: none;}
    ul.postlist-style-circle {list-style-type: circle;}
  </style>

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <blockquote>
<div><table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Author:</th><td class="field-body">aragaven</td>
</tr>
</tbody>
</table>
</div></blockquote>
<div class="section" id="bioflows-tutorials">
<h1>2 BioFlows Tutorials<a class="headerlink" href="#bioflows-tutorials" title="Permalink to this headline">¶</a></h1>
<p>This section provides an overview of how to run pre-defined workflows using
the <strong>bioflows</strong> package. The tutorials are based on presuming that the analysis is
conducted on the Brown <a class="reference external" href="https://web1.ccv.brown.edu">CCV</a> compute cluster, to use on other systems please
check the installation instructions for setting up your environment. Currently, we have implemented an RNA-seq
workflow using the GSNAP RNAseq aligner. This will be updated as new
workflows and enhnacements are made to the BioFlows package.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last"><em>If you are not sure what the console means or how to login to CCV go to this resource.</em></p>
</div>
<div class="section" id="setup-the-environment-for-bioflows">
<h2>2.1 Setup the Environment for bioflows<a class="headerlink" href="#setup-the-environment-for-bioflows" title="Permalink to this headline">¶</a></h2>
<p>First make sure the conda environment is setup in your PATH
variable. In your CCV console type</p>
<div class="code sh highlight-default notranslate"><div class="highlight"><pre><span></span>echo $PATH
</pre></div>
</div>
<p>and you should see <code class="docutils literal notranslate"><span class="pre">/gpfs/data/cbc/cbc_conda_v1/bin/</span></code> as the first
element in the list of paths in your output.</p>
<div class="admonition warning">
<p class="first admonition-title">Warning</p>
<dl class="docutils">
<dt>If  <code class="docutils literal notranslate"><span class="pre">/gpfs/data/cbc/cbc_conda_v1/bin/</span></code> is not the first element in the</dt>
<dd>list then use the command</dd>
</dl>
<div class="code sh highlight-default notranslate"><div class="highlight"><pre><span></span>export PATH=/gpfs/data/cbc/cbc_conda_v1/bin/:$PATH
</pre></div>
</div>
<p class="last">This will add  <code class="docutils literal notranslate"><span class="pre">/gpfs/data/cbc/cbc_conda_v1/bin/</span></code> to the beginning of
your <code class="docutils literal notranslate"><span class="pre">PATH</span></code> variable</p>
</div>
<p>For convenience we will use <code class="docutils literal notranslate"><span class="pre">/users/username</span></code> as the working directory and you should modify
accordingly.</p>
<p>These scripts should be run in a persistent terminal session and we will
use GNU screen to do that, so that the we can disconnnect from our ssh
sessions for long running jobs. To learn more on how to use screen use
the following link
<a class="reference external" href="https://www.linode.com/docs/networking/ssh/using-gnu-screen-to-manage-persistent-terminal-sessions">gnu
screen tutorial</a></p>
<p>Start a screen session naming it test_bioflows as below</p>
<div class="code sh highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">screen</span> <span class="o">-</span><span class="n">S</span> <span class="n">test_bioflows</span>
</pre></div>
</div>
<p>once you are in your screen session you set up your python environment
with the following commands</p>
<div class="code sh highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">source</span> <span class="n">activate</span> <span class="n">bflows</span>
</pre></div>
</div>
<p>Now you are ready to run the predefined RNAseq workflow</p>
</div>
<div class="section" id="tutorial-1-rnaseq-with-gsnap">
<h2>2.2 Tutorial 1. RNAseq with GSNAP<a class="headerlink" href="#tutorial-1-rnaseq-with-gsnap" title="Permalink to this headline">¶</a></h2>
<p>This tutorial show how to run a standard predefined RNA-seq analysis on
the using the bioflows tool. This workflow
consists of the following steps which have been implemented:</p>
<ul class="simple">
<li><strong>Fastqc</strong>: For QC of Raw Fastq reads</li>
<li><strong>GSNAP</strong> alignment of the reads to the reference genome of the reads</li>
<li><strong>Qualimap</strong> tool for the QC of the aligments generated</li>
<li><strong>featureCounts</strong> tool for quantifying expression based on mapped reads</li>
<li><strong>Salmon</strong> tool for alignment free quantification of known transcripts</li>
</ul>
<p>Below is a short how-to with all the commands</p>
<div class="section" id="setup-the-yaml-configuration-file">
<h3>2.2.1 Setup the YAML configuration file<a class="headerlink" href="#setup-the-yaml-configuration-file" title="Permalink to this headline">¶</a></h3>
<p>For the test run we use the following YAML format control file. A
detailed documentation of the YAML file and all the options is shown
here. For the current example, we will discuss each section in detail  below</p>
<div class="code yaml highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">bioproject</span><span class="p">:</span> <span class="n">Project_test_localhost</span>
<span class="n">experiment</span><span class="p">:</span> <span class="p">:</span><span class="n">red</span><span class="p">:</span><span class="n">rnaseq_pilot</span>
<span class="n">sample_manifest</span><span class="p">:</span>
  <span class="n">fastq_file</span><span class="p">:</span> <span class="n">sampl_manifest_min</span><span class="o">.</span><span class="n">csv</span>
  <span class="n">metadata</span><span class="p">:</span>
<span class="n">run_parms</span><span class="p">:</span>
  <span class="n">conda_command</span><span class="p">:</span> <span class="n">source</span> <span class="n">activate</span> <span class="o">/</span><span class="n">gpfs</span><span class="o">/</span><span class="n">runtime</span><span class="o">/</span><span class="n">opt</span><span class="o">/</span><span class="n">conda</span><span class="o">/</span><span class="n">envs</span><span class="o">/</span><span class="n">cbc_conda_test</span>
  <span class="n">work_dir</span><span class="p">:</span> <span class="o">/</span><span class="n">gpfs</span><span class="o">/</span><span class="n">scratch</span><span class="o">/</span><span class="n">aragaven</span><span class="o">/</span><span class="n">test_workflow</span>
  <span class="n">log_dir</span><span class="p">:</span> <span class="n">logs</span>
  <span class="n">paired_end</span><span class="p">:</span> <span class="kc">False</span>
  <span class="n">local_targets</span><span class="p">:</span> <span class="kc">False</span>
  <span class="n">db</span><span class="p">:</span> <span class="n">sqlite</span>
  <span class="n">db_loc</span><span class="p">:</span> <span class="s1">&#39;:memory:&#39;</span>
  <span class="n">saga_host</span><span class="p">:</span> <span class="n">localhost</span>
  <span class="n">ssh_user</span><span class="p">:</span> <span class="n">aragaven</span>
  <span class="n">saga_scheduler</span><span class="p">:</span> <span class="n">slurm</span>
  <span class="n">gtf_file</span><span class="p">:</span> <span class="o">/</span><span class="n">gpfs</span><span class="o">/</span><span class="n">data</span><span class="o">/</span><span class="n">cbc</span><span class="o">/</span><span class="n">cbcollab</span><span class="o">/</span><span class="n">ref_tools</span><span class="o">/</span><span class="n">Ensembl_mus_GRCm38</span><span class="o">.</span><span class="n">p5_rel89</span><span class="o">/</span><span class="n">Mus_musculus</span><span class="o">.</span><span class="n">GRCm38</span><span class="o">.</span><span class="mf">89.</span><span class="n">gtf</span>
<span class="n">workflow_sequence</span><span class="p">:</span>
  <span class="n">fastqc</span><span class="p">:</span> <span class="n">default</span>
  <span class="n">gsnap</span><span class="p">:</span>
    <span class="n">options</span><span class="p">:</span>
      <span class="o">-</span><span class="n">d</span><span class="p">:</span> <span class="n">Ensembl_mus_GRCm38</span><span class="o">.</span><span class="n">p5_rel89</span>
      <span class="o">-</span><span class="n">s</span><span class="p">:</span> <span class="n">Mus_musculus</span><span class="o">.</span><span class="n">GRCm38</span><span class="o">.</span><span class="mf">89.</span><span class="n">splicesites</span><span class="o">.</span><span class="n">iit</span>
    <span class="n">job_params</span><span class="p">:</span>
      <span class="n">ncpus</span><span class="p">:</span> <span class="mi">16</span>
      <span class="n">mem</span><span class="p">:</span> <span class="mi">40000</span>
      <span class="n">time</span><span class="p">:</span> <span class="mi">60</span>
  <span class="n">qualimap_rnaseq</span><span class="p">:</span> <span class="n">default</span>
  <span class="n">htseq</span><span class="o">-</span><span class="n">count</span><span class="p">:</span> <span class="n">default</span>
</pre></div>
</div>
</div>
<div class="section" id="yaml-description">
<h3>2.2.2 YAML description<a class="headerlink" href="#yaml-description" title="Permalink to this headline">¶</a></h3>
<dl class="docutils">
<dt>bioproject</dt>
<dd>This will be an unique identifier for your project.
This is adopted from the NCBI SRA format structure, so if you use an
SRA dataset you can employ these ids</dd>
<dt>experiment</dt>
<dd>An identifier for your experiment such as RNA-seq,
ChIP-seq etc</dd>
<dt>sample_manifest</dt>
<dd><p class="first">This section contains two sections</p>
<dl class="last docutils">
<dt>fastq_file</dt>
<dd><p class="first">The full path to the sample to fastq map file.
This file is in a three column comma separated format with each line formatted as:</p>
<p><em>sample_id, path_to_fastq_file_for_read1, path_to_fastq_file_for_read2</em></p>
<p>The <em>sample_id</em> is unique and if you are <strong>using single end data</strong> you just need
to specify one column as shown below:</p>
<p class="last"><em>sample_id, path_to_fastq_file</em></p>
</dd>
<dt>metadata</dt>
<dd>This is all the metadata associated with a given
<em>sample_id</em> if available such as gender, extraction date etc. This
should also be a CSV format file. Currently, not necessary as this
information is not yet used</dd>
</dl>
</dd>
<dt>run_parms</dt>
<dd><p class="first">This section specifies the global parameters for the
current analysis</p>
<dl class="last docutils">
<dt>conda_command</dt>
<dd>This is the command used to activate your conda
environment</dd>
<dt>work_dir</dt>
<dd>The working directory for analysis usually created on  <strong>*gpfs/scratch*</strong></dd>
<dt>log_dir</dt>
<dd>The subdirectory for all the log files</dd>
<dt>paired_end</dt>
<dd>Whether data consists of paired end reads or single end
reads (True/False)</dd>
<dt>local_targets</dt>
<dd>Whether this worklfow is being run from a local
machine</dd>
<dt>db</dt>
<dd>default database engine to use ( sqlite)</dd>
<dt>db_loc</dt>
<dd>location of the database ‘:memory:’</dd>
<dt>saga_host</dt>
<dd>The hostname if workflow is run from a local machine</dd>
<dt>ssh_user</dt>
<dd>The user name if workflow is run from a local machine</dd>
<dt>saga_scheduler</dt>
<dd>The scheduler being used, for CCV the value
used here is <a href="#id1"><span class="problematic" id="id2">:redtext:`/slurm/`</span></a></dd>
<dt>gtf_file</dt>
<dd>The full path to the gtf file for gene annotations</dd>
</dl>
</dd>
<dt>workflow_sequence</dt>
<dd><p class="first">This section specifies the sequence of tools to
be used and the options passed to tools as well as the job parameters
if using a scheduler such as slurm</p>
<dl class="last docutils">
<dt>fastqc</dt>
<dd>If you want to use the default parameters use <em>default</em>
else you can use any of the options provided by the
program. See  the example for GSNAP below on how to do
that. See the documentation for the options for fastqc.</dd>
<dt>gsnap</dt>
<dd><p class="first">Here we give an example of two sections as we need to
pass the index information to the aligner</p>
<dl class="last docutils">
<dt>options</dt>
<dd><p class="first">Specify program options here. In this example we
specify the following</p>
<table class="docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group">
<kbd><span class="option">-d</span></kbd></td>
<td>The genome index for GSNAP</td></tr>
<tr><td class="option-group">
<kbd><span class="option">-s</span></kbd></td>
<td>and the splicesites file location for GSNAP.</td></tr>
</tbody>
</table>
<p>The format is exactly that as to what you would specify on the command line for the program</p>
<div class="code sh highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">-</span><span class="n">d</span> <span class="n">Ensembl_mus_GRCm38</span><span class="o">.</span><span class="n">p5_rel89</span>
<span class="o">-</span><span class="n">s</span> <span class="n">Mus_musculus</span><span class="o">.</span><span class="n">GRCm38</span><span class="o">.</span><span class="mf">89.</span><span class="n">splicesites</span><span class="o">.</span><span class="n">iit</span>
</pre></div>
</div>
<p class="last">See the documentation for the GSNAP program for more options</p>
</dd>
<dt>job_params</dt>
<dd><p class="first">This section specifies parameters for job submission such as memory, number of cores etc</p>
<ul class="last simple">
<li>ncpus: 16</li>
<li>mem: 40000</li>
<li>time: 60</li>
</ul>
</dd>
</dl>
</dd>
<dt>qualimap_rnaseq</dt>
<dd>Run the qualimap module for RNAseq with the <strong>default</strong> settings</dd>
</dl>
</dd>
</dl>
<p>The final YAML control file should look as below to run a test example. Only modify the parts
that are highlighted below to fill in your ownm values.</p>
<div class="code yaml highlight-default notranslate"><div class="highlight"><pre><span></span>bioproject: Project_test_localhost
experiment: rnaseq_pilot
sample_manifest:
  fastq_file: /users/:bluetext:`username/sample_manifest_min.csv`
  metadata:
run_parms:
  conda_command: source activate /gpfs/runtime/opt/conda/envs/cbc_conda_test
  work_dir: /gpfs/scratch/&#39;user/test_workflow&#39;
  log_dir: logs
  paired_end: False
  local_targets: False
  db: sqlite
  db_loc: &quot;:memory:&quot;
  saga_host: localhost
  ssh_user: &#39;ccv username&#39;
  saga_scheduler: slurm
  gtf_file: /gpfs/data/cbc/cbcollab/ref_tools/Ensembl_hg_GRCh37_rel87/Homo_sapiens.GRCh37.87.gtf
workflow_sequence:
  fastqc: default
  gsnap:
    options:
      -d: Ensembl_Homo_sapiens_GRCh37
      -s: Ensembl_Homo_sapiens.GRCh37.87.splicesites.iit
    job_params:
      ncpus: 16
      mem: 40000
    time: 60
  qualimap_rnaseq: default
  htseq-count: default
</pre></div>
</div>
</div>
<div class="section" id="how-to-run">
<h3>2.2.3 How to run<a class="headerlink" href="#how-to-run" title="Permalink to this headline">¶</a></h3>
<p>Copy the above into a text file and save it in <strong>/users/username</strong> as
test_run.yaml</p>
<p>Copy the manifest below into a text file and save it in
<strong>/users/username</strong> as sample_manifest_min.csv</p>
<div class="code sh highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">samp_1299</span><span class="p">,</span><span class="o">/</span><span class="n">gpfs</span><span class="o">/</span><span class="n">scratch</span><span class="o">/</span><span class="n">aragaven</span><span class="o">/</span><span class="n">rnaseq_test</span><span class="o">/</span><span class="n">PE_hg</span><span class="o">/</span><span class="n">Cb2_1</span><span class="o">.</span><span class="n">gz</span><span class="p">,</span><span class="o">/</span><span class="n">gpfs</span><span class="o">/</span><span class="n">scratch</span><span class="o">/</span><span class="n">aragaven</span><span class="o">/</span><span class="n">rnaseq_test</span><span class="o">/</span><span class="n">PE_hg</span><span class="o">/</span><span class="n">Cb2_2</span><span class="o">.</span><span class="n">gz</span>
<span class="n">samp_1214</span><span class="p">,</span><span class="o">/</span><span class="n">gpfs</span><span class="o">/</span><span class="n">scratch</span><span class="o">/</span><span class="n">aragaven</span><span class="o">/</span><span class="n">rnaseq_test</span><span class="o">/</span><span class="n">PE_hg</span><span class="o">/</span><span class="n">Cb_1</span><span class="o">.</span><span class="n">gz</span><span class="p">,</span><span class="o">/</span><span class="n">gpfs</span><span class="o">/</span><span class="n">scratch</span><span class="o">/</span><span class="n">aragaven</span><span class="o">/</span><span class="n">rnaseq_test</span><span class="o">/</span><span class="n">PE_hg</span><span class="o">/</span><span class="n">Cb_2</span><span class="o">.</span><span class="n">gz</span>
</pre></div>
</div>
<p>Now in your screen session run the following commands to setup your
environment if you have not done so previously during the setup or you
have started a new screen session</p>
<div class="code sh highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">source</span> <span class="n">activate</span> <span class="n">bflows</span>
<span class="n">bioflows</span><span class="o">-</span><span class="n">rnaseq</span> <span class="n">test_run</span><span class="o">.</span><span class="n">yaml</span>
</pre></div>
</div>
<p>In this case I have created a small test dataset with 10000 reads from a
test human RNAseq data, so it should run within the hour and you should
see that the alignments are completed.</p>
</div>
</div>
</div>

  <div class="section">
  
  
  </div>

          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="../">bioflows</a></h1>










<h3>Navigation</h3>
<p class="caption"><span class="caption-text">Contents</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../intro/">1 INTRODUCTION</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">2 BioFlows Tutorials</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#setup-the-environment-for-bioflows">2.1 Setup the Environment for bioflows</a></li>
<li class="toctree-l2"><a class="reference internal" href="#tutorial-1-rnaseq-with-gsnap">2.2 Tutorial 1. RNAseq with GSNAP</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../wrappers/">3 Wrappers</a></li>
</ul>


<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../search/" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2018, Ashok Ragavendran.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.7.2</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.10</a>
      
      |
      <a href="../_sources/tutorials.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>