# About **bioflows** Tutorials

This section provides an overview of how to run pre-defined workflows using
the **bioflows** package. The tutorials are based on the presumption that the analysis is
conducted on the Brown [OSCAR](https://web1.ccv.brown.edu) compute cluster To use on other systems please
check the installation instructions for setting up your environment appropriately. Currently, we provide tutorials for implementing an RNA-seq
workflow using the GSNAP RNAseq aligner, Variant Calling with GATK and 16S sequence analysis with QIIME2. This will be updated as new workflows and enhancements are made to the **bioflows** package.

!!! note
    If you are not sure what the console means or how to login to OSCAR go [here]().
