
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>3 Wrappers &#8212; bioflows 0.99 alpha documentation</title>
    <link rel="stylesheet" href="../_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <script type="text/javascript" src="../_static/documentation_options.js"></script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="author" title="About these documents" href="../about/" />
    <link rel="index" title="Index" href="../genindex/" />
    <link rel="search" title="Search" href="../search/" />
    <link rel="prev" title="2 BioFlows Tutorials" href="../tutorials/" />
  
   
  <link rel="stylesheet" href="../_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  
  <link rel="alternate" type="application/atom+xml"  href="../blog/atom.xml" title="bioflows Blog">
  
  
  <link href="True" rel="stylesheet">
  
  <style type="text/css">
    ul.ablog-archive {list-style: none; overflow: auto; margin-left: 0px}
    ul.ablog-archive li {float: left; margin-right: 5px; font-size: 80%}
    ul.postlist a {font-style: italic;}
    ul.postlist-style-disc {list-style-type: disc;}
    ul.postlist-style-none {list-style-type: none;}
    ul.postlist-style-circle {list-style-type: circle;}
  </style>

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="contents topic" id="contents">
<p class="topic-title first">Contents</p>
<ul class="simple">
<li><a class="reference internal" href="#wrappers" id="id2">3 Wrappers</a><ul>
<li><a class="reference internal" href="#fastqc" id="id3">3.1 Fastqc</a></li>
<li><a class="reference internal" href="#gsnap" id="id4">3.2 GSNAP</a></li>
<li><a class="reference internal" href="#qualimap" id="id5">3.3 Qualimap</a></li>
<li><a class="reference internal" href="#htseq-counts" id="id6">3.4 HTSeq Counts</a></li>
</ul>
</li>
</ul>
</div>
<div class="section" id="wrappers">
<h1><a class="toc-backref" href="#id2">3 Wrappers</a><a class="headerlink" href="#wrappers" title="Permalink to this headline">¶</a></h1>
<p>This section provides an overview of all the bioinformatic programs
that currenlty have wrappers implemented internally in BioFlows. For
each tool all the options are shown and any of these options can be
provided under options in the YAML file.</p>
<div class="section" id="fastqc">
<h2><a class="toc-backref" href="#id3">3.1 Fastqc</a><a class="headerlink" href="#fastqc" title="Permalink to this headline">¶</a></h2>
<p>Fastqc is a java program to undertake QC for fastq reads produced by
NGS, primarily illumina sequencing. More details on the tool as well
as the documentation from the original developers is available from the
<a class="reference external" href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/">Fastqc</a>
website. The following options are
currently available through the fastqc program wrapper in the bioflows package</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>$ fastqc -h

            FastQC - A high throughput sequence QC analysis tool

SYNOPSIS

	fastqc seqfile1 seqfile2 .. seqfileN

    fastqc [-o output dir] [--(no)extract] [-f fastq|bam|sam] 
           [-c contaminant file] seqfile1 .. seqfileN

DESCRIPTION

    FastQC reads a set of sequence files and produces from each one a quality
    control report consisting of a number of different modules, each one of 
    which will help to identify a different potential type of problem in your
    data.
    
    If no files to process are specified on the command line then the program
    will start as an interactive graphical application.  If files are provided
    on the command line then the program will run with no user interaction
    required.  In this mode it is suitable for inclusion into a standardised
    analysis pipeline.
    
    The options for the program as as follows:
    
    -h --help       Print this help file and exit
    
    -v --version    Print the version of the program and exit
    
    -o --outdir     Create all output files in the specified output directory.
                    Please note that this directory must exist as the program
                    will not create it.  If this option is not set then the 
                    output file for each sequence file is created in the same
                    directory as the sequence file which was processed.
                    
    --casava        Files come from raw casava output. Files in the same sample
                    group (differing only by the group number) will be analysed
                    as a set rather than individually. Sequences with the filter
                    flag set in the header will be excluded from the analysis.
                    Files must have the same names given to them by casava
                    (including being gzipped and ending with .gz) otherwise they
                    won&#39;t be grouped together correctly.
                    
    --nano          Files come from naopore sequences and are in fast5 format. In
                    this mode you can pass in directories to process and the program
                    will take in all fast5 files within those directories and produce
                    a single output file from the sequences found in all files.                    
                    
    --nofilter      If running with --casava then don&#39;t remove read flagged by
                    casava as poor quality when performing the QC analysis.
                   
    --extract       If set then the zipped output file will be uncompressed in
                    the same directory after it has been created.  By default
                    this option will be set if fastqc is run in non-interactive
                    mode.
                    
    -j --java       Provides the full path to the java binary you want to use to
                    launch fastqc. If not supplied then java is assumed to be in
                    your path.
                   
    --noextract     Do not uncompress the output file after creating it.  You
                    should set this option if you do not wish to uncompress
                    the output when running in non-interactive mode.
                    
    --nogroup       Disable grouping of bases for reads &gt;50bp. All reports will
                    show data for every base in the read.  WARNING: Using this
                    option will cause fastqc to crash and burn if you use it on
                    really long reads, and your plots may end up a ridiculous size.
                    You have been warned!
                    
    -f --format     Bypasses the normal sequence file format detection and
                    forces the program to use the specified format.  Valid
                    formats are bam,sam,bam_mapped,sam_mapped and fastq
                    
    -t --threads    Specifies the number of files which can be processed
                    simultaneously.  Each thread will be allocated 250MB of
                    memory so you shouldn&#39;t run more threads than your
                    available memory will cope with, and not more than
                    6 threads on a 32 bit machine
                  
    -c              Specifies a non-default file which contains the list of
    --contaminants  contaminants to screen overrepresented sequences against.
                    The file must contain sets of named contaminants in the
                    form name[tab]sequence.  Lines prefixed with a hash will
                    be ignored.

    -a              Specifies a non-default file which contains the list of
    --adapters      adapter sequences which will be explicity searched against
                    the library. The file must contain sets of named adapters
                    in the form name[tab]sequence.  Lines prefixed with a hash
                    will be ignored.
                    
    -l              Specifies a non-default file which contains a set of criteria
    --limits        which will be used to determine the warn/error limits for the
                    various modules.  This file can also be used to selectively 
                    remove some modules from the output all together.  The format
                    needs to mirror the default limits.txt file found in the
                    Configuration folder.
                    
   -k --kmers       Specifies the length of Kmer to look for in the Kmer content
                    module. Specified Kmer length must be between 2 and 10. Default
                    length is 7 if not specified.
                    
   -q --quiet       Supress all progress messages on stdout and only report errors.
   
   -d --dir         Selects a directory to be used for temporary files written when
                    generating report images. Defaults to system temp directory if
                    not specified.
                    
BUGS

    Any bugs in fastqc should be reported either to simon.andrews@babraham.ac.uk
    or in www.bioinformatics.babraham.ac.uk/bugzilla/
</pre></div>
</div>
</div>
<div class="section" id="gsnap">
<h2><a class="toc-backref" href="#id4">3.2 GSNAP</a><a class="headerlink" href="#gsnap" title="Permalink to this headline">¶</a></h2>
<p>GSNAP is a splice aware RNAseq aligner, which is consistently among
the top aligners in the DREAM challenge (ref required).</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>$ gsnap --help
Note: gsnap.avx2 does not exist.  For faster speed, may want to compile package on an AVX2 machine
GSNAP version 2016-09-23 called with args: gsnap.sse42 --help
Usage: gsnap [OPTIONS...] &lt;FASTA file&gt;, or
       cat &lt;FASTA file&gt; | gmap [OPTIONS...]

Input options (must include -d)
  -D, --dir=directory            Genome directory.  Default (as specified by --with-gmapdb to the configure program) is
                                   /Users/aragaven/anaconda2/envs/cbc/share
  -d, --db=STRING                Genome database
  --use-sarray=INT               Whether to use a suffix array, which will give increased speed.
                                   Allowed values: 0 (no), 1 (yes, plus GSNAP/GMAP algorithm, default),
                                   or 2 (yes, and use only suffix array algorithm).
                                   Note that suffix arrays will bias against SNP alleles in
                                   SNP-tolerant alignment.
  -k, --kmer=INT                 kmer size to use in genome database (allowed values: 16 or less)
                                   If not specified, the program will find the highest available
                                   kmer size in the genome database
  --sampling=INT                 Sampling to use in genome database.  If not specified, the program
                                   will find the smallest available sampling value in the genome database
                                   within selected k-mer size
  -q, --part=INT/INT             Process only the i-th out of every n sequences
                                   e.g., 0/100 or 99/100 (useful for distributing jobs
                                   to a computer farm).
  --input-buffer-size=INT        Size of input buffer (program reads this many sequences
                                   at a time for efficiency) (default 1000)
  --barcode-length=INT           Amount of barcode to remove from start of read
                                   (default 0)
  --orientation=STRING           Orientation of paired-end reads
                                   Allowed values: FR (fwd-rev, or typical Illumina; default),
                                   RF (rev-fwd, for circularized inserts), or FF (fwd-fwd, same strand)
  --fastq-id-start=INT           Starting position of identifier in FASTQ header, space-delimited (&gt;= 1)
  --fastq-id-end=INT             Ending position of identifier in FASTQ header, space-delimited (&gt;= 1)
                                 Examples:
                                   @HWUSI-EAS100R:6:73:941:1973#0/1
                                      start=1, end=1 (default) =&gt; identifier is HWUSI-EAS100R:6:73:941:1973#0
                                   @SRR001666.1 071112_SLXA-EAS1_s_7:5:1:817:345 length=36
                                      start=1, end=1  =&gt; identifier is SRR001666.1
                                      start=2, end=2  =&gt; identifier is 071112_SLXA-EAS1_s_7:5:1:817:345
                                      start=1, end=2  =&gt; identifier is SRR001666.1 071112_SLXA-EAS1_s_7:5:1:817:345
  --force-single-end             When multiple FASTQ files are provided on the command line, GSNAP assumes
                                    they are matching paired-end files.  This flag treats each file as single-end.
  --filter-chastity=STRING       Skips reads marked by the Illumina chastity program.  Expecting a string
                                   after the accession having a &#39;Y&#39; after the first colon, like this:
                                         @accession 1:Y:0:CTTGTA
                                   where the &#39;Y&#39; signifies filtering by chastity.
                                   Values: off (default), either, both.  For &#39;either&#39;, a &#39;Y&#39; on either end
                                   of a paired-end read will be filtered.  For &#39;both&#39;, a &#39;Y&#39; is required
                                   on both ends of a paired-end read (or on the only end of a single-end read).
  --allow-pe-name-mismatch       Allows accession names of reads to mismatch in paired-end files
  --gunzip                       Uncompress gzipped input files
  --bunzip2                      Uncompress bzip2-compressed input files

Computation options

  Note: GSNAP has an ultrafast algorithm for calculating mismatches up to and including
((readlength+2)/kmer - 2) (&quot;ultrafast mismatches&quot;).  The program will run fastest if
max-mismatches (plus suboptimal-levels) is within that value.
Also, indels, especially end indels, take longer to compute, although the algorithm
is still designed to be fast.

  -B, --batch=INT                Batch mode (default = 2)
                                 Mode     Offsets       Positions       Genome          Suffix array
                                   0      see note      mmap            mmap            mmap
                                   1      see note      mmap &amp; preload  mmap            mmap
                                   2      see note      mmap &amp; preload  mmap &amp; preload  mmap &amp; preload
                                   3      see note      allocate        mmap &amp; preload  mmap &amp; preload
                      (default)    4      see note      allocate        allocate        mmap &amp; preload
                                   5      see note      allocate        allocate        allocate
                           Note: For a single sequence, all data structures use mmap
                           If mmap not available and allocate not chosen, then will use fileio (very slow)
                       Note about offsets: Expansion of offsets can be controlled
                       independently by the --expand-offsets flag.  However, offsets
                       are accessed relatively fast in this version of GSNAP.

  --use-shared-memory=INT        If 1 (default), then allocated memory is shared among all processes
                                   on this node.  If 0, then each process has private allocated memory
  --preload-shared-memory        Load files indicated by --batch mode into shared memory for use by other
                                   GMAP/GSNAP processes on this node, and then exit.  Ignore any input files.
  --unload-shared-memory         Unload files indicated by --batch mode into shared memory, or allow them
                                   to be unloaded when existing GMAP/GSNAP processes on this node are finished
                                   with them.  Ignore any input files.
  --expand-offsets=INT           Whether to expand the genomic offsets index
                                   Values: 0 (no, default), or 1 (yes).
                                   Expansion gives faster alignment, but requires more memory
  -m, --max-mismatches=FLOAT     Maximum number of mismatches allowed (if not specified, then
                                   defaults to the ultrafast level of ((readlength+index_interval-1)/kmer - 2))
                                   (By default, the genome index interval is 3, but this can be changed
                                    by providing a different value for -q to gmap_build when processing
                                   the genome.)
			                                   If specified between 0.0 and 1.0, then treated as a fraction
                                   of each read length.  Otherwise, treated as an integral number
                                   of mismatches (including indel and splicing penalties)
                                   For RNA-Seq, you may need to increase this value slightly
                                   to align reads extending past the ends of an exon.
  --min-coverage=FLOAT           Minimum coverage required for an alignment.
                                   If specified between 0.0 and 1.0, then treated as a fraction
                                   of each read length.  Otherwise, treated as an integral number
                                   of base pairs.  Default value is 0.0.
  --query-unk-mismatch=INT       Whether to count unknown (N) characters in the query as a mismatch
                                   (0=no (default), 1=yes)
  --genome-unk-mismatch=INT      Whether to count unknown (N) characters in the genome as a mismatch
                                   (0=no, 1=yes (default))
  --maxsearch=INT                Maximum number of alignments to find (default 1000).
                                   Must be larger than --npaths, which is the number to report.
                                   Keeping this number large will allow for random selection among multiple alignments.
                                   Reducing this number can speed up the program.
  -i, --indel-penalty=INT        Penalty for an indel (default 2).
                                   Counts against mismatches allowed.  To find indels, make
                                   indel-penalty less than or equal to max-mismatches.
                                   A value &lt; 2 can lead to false positives at read ends
  --indel-endlength=INT          Minimum length at end required for indel alignments (default 4)
  -y, --max-middle-insertions=INT  Maximum number of middle insertions allowed (default is readlength - indel-endlength)
  -z, --max-middle-deletions=INT Maximum number of middle deletions allowed (default 30)
  -Y, --max-end-insertions=INT   Maximum number of end insertions allowed (default 3)
  -Z, --max-end-deletions=INT    Maximum number of end deletions allowed (default 6)
  -M, --suboptimal-levels=INT    Report suboptimal hits beyond best hit (default 0)
                                   All hits with best score plus suboptimal-levels are reported
  -a, --adapter-strip=STRING     Method for removing adapters from reads.  Currently allowed values: off, paired.
                                   Default is &quot;off&quot;.  To turn on, specify &quot;paired&quot;, which removes adapters
                                   from paired-end reads if they appear to be present.
  --trim-mismatch-score=INT      Score to use for mismatches when trimming at ends (default is -3;
                                   to turn off trimming, specify 0).  Warning: turning trimming off
                                   will give false positive mismatches at the ends of reads
  --trim-indel-score=INT         Score to use for indels when trimming at ends (default is -2;
                                   to turn off trimming, specify 0).  Warning: turning trimming off
                                   will give false positive indels at the ends of reads
  --end-detail=STRING            Amount of alignment detail at ends of read: high, medium, or low (default)
                                   Note: medium detail could increase speed by 20% or so, but will miss some
                                   splices at the ends of reads
  -V, --snpsdir=STRING           Directory for SNPs index files (created using snpindex) (default is
                                   location of genome index files specified using -D and -d)
   -v, --use-snps=STRING          Use database containing known SNPs (in &lt;STRING&gt;.iit, built
                                   previously using snpindex) for tolerance to SNPs
  --cmetdir=STRING               Directory for methylcytosine index files (created using cmetindex)
                                   (default is location of genome index files specified using -D, -V, and -d)
  --atoidir=STRING               Directory for A-to-I RNA editing index files (created using atoiindex)
                                   (default is location of genome index files specified using -D, -V, and -d)
  --mode=STRING                  Alignment mode: standard (default), cmet-stranded, cmet-nonstranded,
                                    atoi-stranded, atoi-nonstranded, ttoc-stranded, or ttoc-nonstranded.
                                    Non-standard modes requires you to have previously run the cmetindex
                                    or atoiindex programs (which also cover the ttoc modes) on the genome
  -t, --nthreads=INT             Number of worker threads
  --max-anchors=INT              Controls number of candidate segments returned by the complete set algorithm
                                   Default is 10.  Can be increased to higher values to solve alignments with
                                   evenly spaced mismatches at close distances.  However, higher values will
                                   cause GSNAP to run more slowly.  A value of 1000, for example, slows down
                                   the program by a factor of 10 or so.  Therefore, change this value only if
                                   absolutely necessary.

Options for GMAP alignment within GSNAP
  --gmap-mode=STRING             Cases to use GMAP for complex alignments containing multiple splices or indels
                                 Allowed values: none, all, pairsearch, indel_knownsplice, terminal, improve
                                   (or multiple values, separated by commas).
                                   Default: all, i.e., pairsearch,indel_knownsplice,terminal,improve
  --trigger-score-for-gmap=INT   Try GMAP pairsearch on nearby genomic regions if best score (the total
                                   of both ends if paired-end) exceeds this value (default 5)
  --gmap-min-match-length=INT    Keep GMAP hit only if it has this many consecutive matches (default 20)
  --gmap-allowance=INT           Extra mismatch/indel score allowed for GMAP alignments (default 3)
  --max-gmap-pairsearch=INT      Perform GMAP pairsearch on nearby genomic regions up to this many
                                   many candidate ends (default 50).  Requires pairsearch in --gmap-mode
  --max-gmap-terminal=INT        Perform GMAP terminal on nearby genomic regions up to this many
                                   candidate ends (default 50).  Requires terminal in --gmap-mode
  --max-gmap-improvement=INT     Perform GMAP improvement on nearby genomic regions up to this many
                                   candidate ends (default 5).  Requires improve in --gmap-mode
  --microexon-spliceprob=FLOAT   Allow microexons only if one of the splice site probabilities is
                                   greater than this value (default 0.95)

Splicing options for DNA-Seq
  --find-dna-chimeras=INT              Look for distant splicing in DNA-Seq data (0=no (default), 1=yes)
                                         Automatically inactivated for RNA-Seq data
                                         if -N or -s are specified)

Splicing options for RNA-Seq
  -N, --novelsplicing=INT              Look for novel splicing (0=no (default), 1=yes)
  --splicingdir=STRING                 Directory for splicing involving known sites or known introns,
                                         as specified by the -s or --use-splicing flag (default is
                                         directory computed from -D and -d flags).  Note: can
                                         just give full pathname to the -s flag instead.
  -s, --use-splicing=STRING            Look for splicing involving known sites or known introns
                                         (in &lt;STRING&gt;.iit), at short or long distances
                                         See README instructions for the distinction between known sites
                                         and known introns
  --ambig-splice-noclip                For ambiguous known splicing at ends of the read, do not clip at the
                                         splice site, but extend instead into the intron.  This flag makes
                                         sense only if you provide the --use-splicing flag, and you are trying
                                         to eliminate all soft clipping with --trim-mismatch-score=0
  -w, --localsplicedist=INT            Definition of local novel splicing event (default 200000)
  --novelend-splicedist=INT            Distance to look for novel splices at the ends of reads (default 50000)
  -e, --local-splice-penalty=INT       Penalty for a local splice (default 0).  Counts against mismatches allowed
  -E, --distant-splice-penalty=INT     Penalty for a distant splice (default 1).  A distant splice is one where
                                         the intron length exceeds the value of -w, or --localsplicedist, or is an
                                         inversion, scramble, or translocation between two different chromosomes
                                         Counts against mismatches allowed
  -K, --distant-splice-endlength=INT   Minimum length at end required for distant spliced alignments (default 20, min
                                         allowed is the value of -k, or kmer size)
  -l, --shortend-splice-endlength=INT  Minimum length at end required for short-end spliced alignments (default 2,
                                         but unless known splice sites are provided with the -s flag, GSNAP may still
                                         need the end length to be the value of -k, or kmer size to find a given splice
  --distant-splice-identity=FLOAT      Minimum identity at end required for distant spliced alignments (default 0.95)
  --antistranded-penalty=INT           (Not currently implemented, since it leads to poor results)
                                         Penalty for antistranded splicing when using stranded RNA-Seq protocols.
                                         A positive value, such as 1, expects antisense on the first read
                                         and sense on the second read.  Default is 0, which treats sense and antisense
                                         equally well
  --merge-distant-samechr              Report distant splices on the same chromosome as a single splice, if possible.
                                         Will produce a single SAM line instead of two SAM lines, which is also done
                                         for translocations, inversions, and scramble events

Options for paired-end reads
  --pairmax-dna=INT              Max total genomic length for DNA-Seq paired reads, or other reads
                                   without splicing (default 1000).  Used if -N or -s is not specified.
                                   This value is also used for circular chromosomes when splicing in
                                   linear chromosomes is allowed
  --pairmax-rna=INT              Max total genomic length for RNA-Seq paired reads, or other reads
                                   that could have a splice (default 200000).  Used if -N or -s is specified.
                                   Should probably match the value for -w, --localsplicedist.
  --pairexpect=INT               Expected paired-end length, used for calling splices in medial part
                                   of paired-end reads (default 200).  Was turned off in previous versions, but reinstated.
  --pairdev=INT                  Allowable deviation from expected paired-end length, used for
                                   calling splices in medial part of paired-end reads (default 100).
                                   Was turned off in previous versions, but reinstated.

Options for quality scores
  --quality-protocol=STRING      Protocol for input quality scores.  Allowed values:
                                   illumina (ASCII 64-126) (equivalent to -J 64 -j -31)
                                   sanger   (ASCII 33-126) (equivalent to -J 33 -j 0)
                                 Default is sanger (no quality print shift)
                                 SAM output files should have quality scores in sanger protocol

                                 Or you can customize this behavior with these flags:
  -J, --quality-zero-score=INT   FASTQ quality scores are zero at this ASCII value
                                   (default is 33 for sanger protocol; for Illumina, select 64)
  -j, --quality-print-shift=INT  Shift FASTQ quality scores by this amount in output
                                   (default is 0 for sanger protocol; to change Illumina input
                                   to Sanger output, select -31)
Output options
  -n, --npaths=INT               Maximum number of paths to print (default 100).
  -Q, --quiet-if-excessive       If more than maximum number of paths are found,
                                   then nothing is printed.
  -O, --ordered                  Print output in same order as input (relevant
                                   only if there is more than one worker thread)
  --show-refdiff                 For GSNAP output in SNP-tolerant alignment, shows all differences
                                   relative to the reference genome as lower case (otherwise, it shows
                                   all differences relative to both the reference and alternate genome)
  --clip-overlap                 For paired-end reads whose alignments overlap, clip the overlapping region.
  --merge-overlap                For paired-end reads whose alignments overlap, merge the two ends into a single end (beta implementation)
  --print-snps                   Print detailed information about SNPs in reads (works only if -v also selected)
                                   (not fully implemented yet)
  --failsonly                    Print only failed alignments, those with no results
  --nofails                      Exclude printing of failed alignments
  -A, --format=STRING            Another format type, other than default.
                                   Currently implemented: sam, m8 (BLAST tabular format)
  --split-output=STRING          Basename for multiple-file output, separately for nomapping,
                                   halfmapping_uniq, halfmapping_mult, unpaired_uniq, unpaired_mult,
                                   paired_uniq, paired_mult, concordant_uniq, and concordant_mult results
  -o, --output-file=STRING       File name for a single stream of output results.
  --failed-input=STRING          Print completely failed alignments as input FASTA or FASTQ format,
                                    to the given file, appending .1 or .2, for paired-end data.
                                    If the --split-output flag is also given, this file is generated
                                    in addition to the output in the .nomapping file.
  --append-output                When --split-output or --failed-input is given, this flag will append output
                                    to the existing files.  Otherwise, the default is to create new files.
  --order-among-best=STRING      Among alignments tied with the best score, order those alignments in this order.
                                    Allowed values: genomic, random (default)
  --output-buffer-size=INT       Buffer size, in queries, for output thread (default 1000).  When the number
                                   of results to be printed exceeds this size, the worker threads are halted
                                   until the backlog is cleared

Options for SAM output
  --no-sam-headers               Do not print headers beginning with &#39;@&#39;
  --add-paired-nomappers         Add nomapper lines as needed to make all paired-end results alternate
                                   between first end and second end
  --paired-flag-means-concordant=INT  Whether the paired bit in the SAM flags means concordant only (1)
                                 or paired plus concordant (0, default)
  --sam-headers-batch=INT        Print headers only for this batch, as specified by -q
  --sam-use-0M                   Insert 0M in CIGAR between adjacent insertions and deletions
                                   Required by Picard, but can cause errors in other tools
  --sam-multiple-primaries       Allows multiple alignments to be marked as primary if they
                                   have equally good mapping scores
  --force-xs-dir                 For RNA-Seq alignments, disallows XS:A:? when the sense direction
                                   is unclear, and replaces this value arbitrarily with XS:A:+.
                                   May be useful for some programs, such as Cufflinks, that cannot
                                   handle XS:A:?.  However, if you use this flag, the reported value
                                   of XS:A:+ in these cases will not be meaningful.
  --md-lowercase-snp             In MD string, when known SNPs are given by the -v flag,
                                   prints difference nucleotides as lower-case when they,
                                   differ from reference but match a known alternate allele
  --extend-soft-clips            Extends alignments through soft clipped regions
  --action-if-cigar-error        Action to take if there is a disagreement between CIGAR length and sequence length
                                   Allowed values: ignore, warning, noprint (default), abort
  --read-group-id=STRING         Value to put into read-group id (RG-ID) field
  --read-group-name=STRING       Value to put into read-group name (RG-SM) field
  --read-group-library=STRING    Value to put into read-group library (RG-LB) field
  --read-group-platform=STRING   Value to put into read-group library (RG-PL) field

Help options
  --check                        Check compiler assumptions
  --version                      Show version
  --help                         Show this help message
</pre></div>
</div>
</div>
<div class="section" id="qualimap">
<h2><a class="toc-backref" href="#id5">3.3 Qualimap</a><a class="headerlink" href="#qualimap" title="Permalink to this headline">¶</a></h2>
<p>Qualimap is the suite of tools for generating metrics associated with
alginment of fastq reads. This tool provides multiple modules
including a module for collecting metrics to RNA-seq alignments. We
use the rnaseq module for generating metrics towards QC of the
alignments.</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>$ qualimap rnaseq --help
Command line run, unsetting DISPLAY variable...
Display: 
Java memory size is set to 1200M
Launching application...

QualiMap v.2.2.2-dev
Built on 2016-12-11 14:41

Selected tool: rnaseq

ERROR: Unrecognized option: --help

usage: qualimap rnaseq [-a &lt;arg&gt;] -bam &lt;arg&gt; -gtf &lt;arg&gt; [-oc &lt;arg&gt;] [-outdir
       &lt;arg&gt;] [-outfile &lt;arg&gt;] [-outformat &lt;arg&gt;] [-p &lt;arg&gt;] [-pe] [-s]
 -a,--algorithm &lt;arg&gt;             Counting algorithm:
                                  uniquely-mapped-reads(default) or
                                  proportional.
 -bam &lt;arg&gt;                       Input mapping file in BAM format.
 -gtf &lt;arg&gt;                       Annotations file in Ensembl GTF format.
 -oc &lt;arg&gt;                        Output file for computed counts. If only name
                                  of the file is provided, then the file will be
                                  saved in the output folder.
 -outdir &lt;arg&gt;                    Output folder for HTML report and raw data.
 -outfile &lt;arg&gt;                   Output file for PDF report (default value is
                                  report.pdf).
 -outformat &lt;arg&gt;                 Format of the output report (PDF, HTML or both
                                  PDF:HTML, default is HTML).
 -p,--sequencing-protocol &lt;arg&gt;   Sequencing library protocol:
                                  strand-specific-forward,
                                  strand-specific-reverse or non-strand-specific
                                  (default)
 -pe,--paired                     Setting this flag for paired-end experiments
                                  will result in counting fragments instead of
                                  reads
 -s,--sorted                      This flag indicates that the input file is
                                  already sorted by name. If not set, additional
                                  sorting by name will be performed. Only
                                  required for paired-end analysis.



Special arguments: 

    --java-mem-size  Use this argument to set Java memory heap size. Example:
                     qualimap bamqc -bam very_large_alignment.bam --java-mem-size=4G
</pre></div>
</div>
</div>
<div class="section" id="htseq-counts">
<h2><a class="toc-backref" href="#id6">3.4 HTSeq Counts</a><a class="headerlink" href="#htseq-counts" title="Permalink to this headline">¶</a></h2>
<p>This script takes one or more alignment files in SAM/BAM format and a feature
file in GFF format and calculates for each feature the number of reads mapping
to it. See <a class="reference external" href="http://htseq.readthedocs.io/en/master/count.html">the htseq docs</a> for
details.</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>$ htseq-count -h
usage: htseq-count [options] alignment_file gff_file

This script takes one or more alignment files in SAM/BAM format and a feature file in GFF format and calculates for each feature the number of reads mapping to it. See
http://htseq.readthedocs.io/en/master/count.html for details.

positional arguments:
  samfilenames          Path to the SAM/BAM files containing the mapped reads. If &#39;-&#39; is selected, read from standard input
  featuresfilename      Path to the file containing the features

optional arguments:
  -h, --help            show this help message and exit
  -f {sam,bam}, --format {sam,bam}
                        type of &lt;alignment_file&gt; data, either &#39;sam&#39; or &#39;bam&#39; (default: sam)
  -r {pos,name}, --order {pos,name}
                        &#39;pos&#39; or &#39;name&#39;. Sorting order of &lt;alignment_file&gt; (default: name). Paired-end sequencing data must be sorted either by position or by read name, and the sorting order must be specified.
                        Ignored for single-end data.
  --max-reads-in-buffer MAX_BUFFER_SIZE
                        When &lt;alignment_file&gt; is paired end sorted by position, allow only so many reads to stay in memory until the mates are found (raising this number will use more memory). Has no effect for
                        single end or paired end sorted by name
  -s {yes,no,reverse}, --stranded {yes,no,reverse}
                        whether the data is from a strand-specific assay. Specify &#39;yes&#39;, &#39;no&#39;, or &#39;reverse&#39; (default: yes). &#39;reverse&#39; means &#39;yes&#39; with reversed strand interpretation
  -a MINAQUAL, --minaqual MINAQUAL
                        skip all reads with alignment quality lower than the given minimum value (default: 10)
  -t FEATURETYPE, --type FEATURETYPE
                        feature type (3rd column in GFF file) to be used, all features of other type are ignored (default, suitable for Ensembl GTF files: exon)
  -i IDATTR, --idattr IDATTR
                        GFF attribute to be used as feature ID (default, suitable for Ensembl GTF files: gene_id)
  --additional-attr ADDITIONAL_ATTR [ADDITIONAL_ATTR ...]
                        Additional feature attributes (default: none, suitable for Ensembl GTF files: gene_name)
  -m {union,intersection-strict,intersection-nonempty}, --mode {union,intersection-strict,intersection-nonempty}
                        mode to handle reads overlapping more than one feature (choices: union, intersection-strict, intersection-nonempty; default: union)
  --nonunique {none,all}
                        Whether to score reads that are not uniquely aligned or ambiguously assigned to features
  --secondary-alignments {score,ignore}
                        Whether to score secondary alignments (0x100 flag)
  --supplementary-alignments {score,ignore}
                        Whether to score supplementary alignments (0x800 flag)
  -o SAMOUTS [SAMOUTS ...], --samout SAMOUTS [SAMOUTS ...]
                        write out all SAM alignment records into an output SAM file called SAMOUT, annotating each line with its feature assignment (as an optional field with tag &#39;XF&#39;)
  -q, --quiet           suppress progress report

Written by Simon Anders (sanders@fs.tum.de), European Molecular Biology Laboratory (EMBL). (c) 2010. Released under the terms of the GNU General Public License v3. Part of the &#39;HTSeq&#39; framework, version 0.9.1.
</pre></div>
</div>
</div>
</div>

  <div class="section">
  
  
  </div>

          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="../">bioflows</a></h1>










<h3>Navigation</h3>
<p class="caption"><span class="caption-text">Contents</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../intro/">1 INTRODUCTION</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/">2 BioFlows Tutorials</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">3 Wrappers</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#fastqc">3.1 Fastqc</a></li>
<li class="toctree-l2"><a class="reference internal" href="#gsnap">3.2 GSNAP</a></li>
<li class="toctree-l2"><a class="reference internal" href="#qualimap">3.3 Qualimap</a></li>
<li class="toctree-l2"><a class="reference internal" href="#htseq-counts">3.4 HTSeq Counts</a></li>
</ul>
</li>
</ul>


<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../search/" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2018, Ashok Ragavendran.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.7.2</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.10</a>
      
      |
      <a href="../_sources/wrappers.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>